### PR TITLE
Apply instant execution system property read detection to Groovy DSL build scripts and Groovy plugins

### PR DESCRIPTION
**Pull request by adammurdoch**
### Context

Improves the instrumentation to detect calls to `System.getProperty(String)` from dynamic Groovy, and also applies the instrumentation to Groovy DSL build scripts.

Also rework the Groovy DSL script caching so that the class name remapping and instrumentation can happen as a single cached step, rather than adding another layer of caching.

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/FileCacheBackedScriptClassCompiler.java [hunk 1]
```diff
@@ -97,32 +101,66 @@ public FileCacheBackedScriptClassCompiler(CacheRepository cacheRepository, Scrip
         // Both caches can be closed directly after use because:
         // For 1, if the script changes or its compile classpath changes, a different directory will be used
         // For 2, if the script changes, a different cache is used. If the classpath changes, the cache is invalidated, but classes are remapped to 1. anyway so never directly used
-        PersistentCache remappedClassesCache = cacheRepository.cache("scripts-remapped/" + source.getClassName() + "/" + sourceHash + "/" + classpathHash)
-            .withDisplayName(dslId + " remapped class cache for " + sourceHash)
-            .withInitializer(new ProgressReportingInitializer(progressLoggerFactory, new RemapBuildScriptsAction<>(remapped, classpathHash, sourceHash, dslId, classLoader, operation, verifier, scriptBaseClass),
+        final PersistentCache cache = cacheRepository.cache("scripts/" + sourceHash + "/" + classpathHash)
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/FileCacheBackedScriptClassCompiler.java:104]: ❓ With this change, shouldn't the script cache cleanup for CI be updated?

https://github.com/gradle/gradle/blob/b5aa8554d5da43d4267ce1d240aa7a98399f23eb/buildSrc/subprojects/cleanup/src/main/kotlin/org/gradle/gradlebuild/testing/integrationtests/cleanup/Cleanup.kt#L95

File: subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java [hunk 2]
```diff
@@ -43,4 +53,19 @@ public static String systemProperty(String key, String consumer) {
     public interface Listener {
         void systemPropertyQueried(String key, String consumer);
     }
+
+    private static class SystemPropertyCallSite extends AbstractCallSite {
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java:57]: Nice!

**Comment by technoir42** [comment]
@adammurdoch I found using `git bisect` that this PR introduced the regression described in #14927. Any advice how we could work around it?

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/FileCacheBackedScriptClassCompiler.java [hunk 3]
```diff
@@ -97,32 +101,66 @@ public FileCacheBackedScriptClassCompiler(CacheRepository cacheRepository, Scrip
...
+                "Compiling " + source.getDisplayName()))
             .open();
         try {
-            File remappedClassesDir = classesDir(remappedClassesCache);
-            File remappedMetadataDir = metadataDir(remappedClassesCache);
-
-            return scriptCompilationHandler.loadFromDir(source, sourceHashCode, targetScope, remappedClassesDir, remappedMetadataDir, operation, scriptBaseClass);
+            File genericClassesDir = classesDir(cache, operation);
+            File metadataDir = metadataDir(cache);
+            File remappedClassesDir = remapClasses(genericClassesDir, remapped);
```
Comment (technoir42) [inline subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/FileCacheBackedScriptClassCompiler.java:115]: @adammurdoch this line takes ~400ms to execute when 2 independent Gradle projects are being configured in parallel but normally it takes ~1ms.

**Pull request merged**